File: backend/src/db/submission.js
const db = require('.');
const DbError = require('../errors/DbError');
const DbConflictError = require('../errors/DbConflictError');
const DbNotFoundError = require('../errors/DbNotFoundError');
const { courseRole } = require('../consts');

const assignmentTable = 'assignment';
const submissionTable = 'submission';
const userCourseTable = 'user_course';

const mapDbSubmission = submission => ({
  id: submission.id,
  result: submission.result,
  submissionStatusName: submission.submission_status_name,
  submitterEmail: submission.submitter_email,
  assignmentId: submission.assignment_id
});

const createSubmission = async submission => {
  const userCourseQuery = `SELECT * FROM ${userCourseTable} WHERE user_email=$1 AND course_id=(SELECT course_id FROM ${assignmentTable} WHERE id=$2)`;
  const userCourseValues = [submission.submitterEmail, submission.assignmentId];

  const createSubmissionQuery = `INSERT INTO ${submissionTable} (id, result, submission_status_name, submitter_email, assignment_id)
  VALUES ($1, $2, $3, $4, $5)
  RETURNING *`;
  const createSubmissionValues = [
    submission.id,
    '{}',
    submission.submissionStatusName,
    submission.submitterEmail,
    submission.assignmentId
  ];

  try {
    let result;
    await db.transcation(async client => {
      result = await client.query(userCourseQuery, userCourseValues);
      if (result.rowCount === 0) {
        throw new Error('user is not in the course');
      }
      result = await db.query(createSubmissionQuery, createSubmissionValues);
    });
    return mapDbSubmission(result.rows[0]);
  } catch (error) {
    const errorMessage = `failed to create submission for assignment with id '${submission.assignmentId}' in the database: ${error.message}`;
    if (db.isConflictError(error)) {
      throw new DbConflictError(errorMessage);
    }
    throw new DbError(errorMessage);
  }
};

const getSubmissions = async (email, assignmentId) => {
  const query = `SELECT * FROM ${submissionTable}
  WHERE assignment_id=$1 AND (submitter_email=$2 OR assignment_id IN (SELECT id FROM ${assignmentTable}
                         WHERE course_id IN (SELECT course_id FROM ${userCourseTable} WHERE user_email=$2)))`;
  const values = [assignmentId, email];

  try {
    const result = await db.query(query, values);
    return result.rows.map(row => mapDbSubmission(row));
  } catch (error) {
    const errorMessage = `failed to find submission for assignment '${assignmentId}' with submitter '${email}' in the database: ${error.message}`;
    throw new DbError(errorMessage);
  }
};

const getSubmission = async (id, email) => {
  const query = `SELECT * FROM ${submissionTable}
  WHERE id=$1 AND (submitter_email=$2 OR assignment_id IN (SELECT id FROM ${assignmentTable}
                       WHERE course_id IN (SELECT course_id FROM ${userCourseTable} WHERE user_email=$2 AND course_role_name='${courseRole.ASSISTANT}')))`;
  const values = [id, email];

  let result;
  try {
    result = await db.query(query, values);
  } catch (error) {
    const errorMessage = `failed to find submission with id '${id}' in the database: ${error.message}`;
    throw new DbError(errorMessage);
  }

  if (result.rowCount === 0) {
    throw new DbNotFoundError(`failed to find submission with id '${id}' in the database`);
  }

  return mapDbSubmission(result.rows[0]);
};

const updateSubmission = () => {};

module.exports = {
  createSubmission,
  getSubmissions,
  getSubmission,
  updateSubmission
};

File: pkg/storage/submission.go
package storage

import (
	"context"
	"fmt"

	"github.com/ZdravkoGyurov/grader/pkg/errors"
	"github.com/ZdravkoGyurov/grader/pkg/types"

	"github.com/jackc/pgx/v4"
)

var (
	readUserFromCourseQuery = fmt.Sprintf(`SELECT * FROM %s WHERE user_email=$1 AND 
	course_id=(SELECT course_id FROM %s WHERE id=$2)`, userCourseTable, assignmentTable)

	insertSubmissionQuery = fmt.Sprintf(`INSERT INTO %s 
	(id, result, submission_status_name, submitter_email, assignment_id)
	VALUES ($1, $2, $3, $4, $5)
	RETURNING *`, submissionTable)

	readSubmissionsQuery = fmt.Sprintf(`SELECT * FROM %s WHERE assignment_id=$1 AND 
	(submitter_email=$2 OR assignment_id IN (SELECT id FROM %s WHERE course_id IN 
		(SELECT course_id FROM %s WHERE user_email=$2)))`,
		submissionTable, assignmentTable, userCourseTable)

	readSubmissionQuery = fmt.Sprintf(`SELECT * FROM %s WHERE id=$1 AND 
	(submitter_email=$2 OR assignment_id IN (SELECT id FROM %s WHERE course_id IN 
	(SELECT course_id FROM %s WHERE user_email=$2 AND course_role_name='%s')))`,
		submissionTable, assignmentTable, userCourseTable, types.CourseRoleAssistant)
)

func (s *Storage) CreateSubmission(ctx context.Context, submission *types.Submission) error {
	dbCtx, cancel := context.WithTimeout(ctx, s.cfg.RequestTimeout)
	defer cancel()

	err := inTransaction(ctx, s.pool, func(tx pgx.Tx) error {
		row := tx.QueryRow(dbCtx, readUserFromCourseQuery, submission.SubmitterEmail, submission.AssignmentID)
		if _, err := readUserCourseRecord(row); err != nil {
			return dbError(errors.Newf("user is not in the course: %w", err))
		}
		if _, err := tx.Exec(dbCtx, insertSubmissionQuery, submission.Fields()...); err != nil {
			return dbError(errors.Newf("failed to create submission: %w", err))
		}
		return nil
	})

	return err
}

func (s *Storage) GetSubmissions(ctx context.Context, userEmail, assignmentID string) ([]*types.Submission, error) {
	dbCtx, cancel := context.WithTimeout(ctx, s.cfg.RequestTimeout)
	defer cancel()

	rows, err := s.pool.Query(dbCtx, readSubmissionsQuery, assignmentID, userEmail)
	if err != nil {
		return nil, dbError(errors.Newf("failed to get submissions: %w", err))
	}

	submissions := make([]*types.Submission, 0)
	for rows.Next() {
		if err := rows.Err(); err != nil {
			return nil, dbError(errors.Newf("failed to read submission row: %w", err))
		}
		submission, err := readSubmissionRecord(rows)
		if err != nil {
			return nil, dbError(errors.Newf("failed to deserialize submission row: %w", err))
		}
		submissions = append(submissions, submission)
	}

	return submissions, nil
}

func (s *Storage) GetSubmission(ctx context.Context, id, userEmail string) (*types.Submission, error) {
	dbCtx, cancel := context.WithTimeout(ctx, s.cfg.RequestTimeout)
	defer cancel()

	row := s.pool.QueryRow(dbCtx, readSubmissionQuery, id, userEmail)
	submission, err := readSubmissionRecord(row)
	if err != nil {
		return nil, dbError(errors.Newf("failed to get submission with id %s: %w", id, err))
	}

	return submission, nil
}

func readSubmissionRecord(row dbRecord) (*types.Submission, error) {
	var submission types.Submission

	err := row.Scan(
		&submission.ID,
		&submission.Result,
		&submission.SubmissionStatusName,
		&submission.SubmitterEmail,
		&submission.AssignmentID,
	)
	if err != nil {
		return nil, err
	}

	return &submission, nil
}
